Service/v2/StorageService.php

```diff
@@ -0,0 +1,156 @@
+
+namespace AppVerk\GoogleCloudStorageMediaBundle\Service\v2;
+
+use AppVerk\GoogleCloudStorageMediaBundle\Dto\UploadedFileDto;
+use AppVerk\GoogleCloudStorageMediaBundle\Exception\FilesystemException as AppFileSystemException;
+use AppVerk\GoogleCloudStorageMediaBundle\Exception\InvalidMimetypeException;
+use AppVerk\GoogleCloudStorageMediaBundle\Exception\InvalidSizeException;
+use AppVerk\GoogleCloudStorageMediaBundle\Flysystem\Retriever\UrlRetrieverInterface;
+use AppVerk\GoogleCloudStorageMediaBundle\Namer\AbstractNamer;
+use AppVerk\GoogleCloudStorageMediaBundle\Service\MediaValidation;
+use League\Flysystem\FilesystemException;
+use League\Flysystem\FilesystemOperator;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class StorageService
+{
+    private FilesystemOperator $filesystem;
+
+    private AbstractNamer $namer;
+
+    private TranslatorInterface $translator;
+
+    private MediaValidation $mediaValidation;
+
+    private UrlRetrieverInterface $urlRetriever;
+
+    public function __construct(
+        MediaValidation $mediaValidation,
+        AbstractNamer $namer,
+        FilesystemOperator $filesystem,
+        TranslatorInterface $translator,
+        UrlRetrieverInterface $urlRetriever
+    )
+    {
+        $this->mediaValidation = $mediaValidation;
+        $this->namer = $namer;
+        $this->filesystem = $filesystem;
+        $this->translator = $translator;
+        $this->urlRetriever = $urlRetriever;
+    }
+
+    public function save(UploadedFile $file, ?string $originalFilename = '', ?string $groupName = null): UploadedFileDto
+    {
+        $this->validate($file, $groupName);
+        $this->validateSize($file, $groupName);
+
+        $url = $filename = $this->namer->generate(
+            $originalFilename ?: $file->getFilename(),
+            $file->getExtension() ?: $file->guessExtension()
+        );
+
+        $splitName = explode('/', $filename);
+        $friendlyName = end($splitName);
+
+        try {
+            $this->filesystem->write($filename, $file->getContent());
+            $url = $this->urlRetriever->getUrl($filename);
+        } catch (FilesystemException $e) {
+            throw new AppFileSystemException($e->getMessage());
+        }
+
+        return new UploadedFileDto(
+            $friendlyName,
+            $filename,
+            $url,
+            $file->getMimeType(),
+            $file->getSize()
+        );
+    }
+
+    public function sanitize(string $filename): string
+    {
+        return preg_replace(['/\s/', '/\.[\.]+/', '/[^\w_\.\-]/'], ['_', '.', ''], $filename);
+    }
+
+    public function remove(string $filename): bool
+    {
+        try {
+            if (true === $this->filesystem->fileExists($filename)) {
+                $this->filesystem->delete($filename);
```

Comment:
Lets add new events for add, remove. Will be needed for example for external cropping/resizing.

Reply:
Added